### DISCUSSION TITLE
No punctuation for the first 75 minutes of the video. What could be the error / bug?

### DISCUSSION THREAD

**Discussion by FurkanGozukara** [openai/whisper#194] [Q&A] [answered]
Hello. I am generating subtitles for my this video : https://www.youtube.com/watch?v=77iDUQd4x90

I have provided the video file directly to the wisher with language en and model large

However as can be seen from the below screenshots, there is no punctuation in the transcription for first 75 minutes of the video : 

But after that, the punctuation starts

I am using the latest updated version of Whisper on Windows computer with Python Version 3.9.9

![image](https://user-images.githubusercontent.com/19240467/193139994-210dd42e-549b-43d1-a4f1-0f56069e72fe.png)

![image](https://user-images.githubusercontent.com/19240467/193140120-11d72d60-d05f-4c0a-bffd-d3022d1337e2.png)

**Comment by FurkanGozukara** [comment]
@jongwook @drdaxxy @cool-RR @gglanzani If you could check I would appreciate very much

**Comment by jongwook** [comment, marked as answer]
Being an autoregressive model, Whisper has a certain chance to get stuck into a "no-punctuation mode". It also seems to be correlated with the tendency to create either precise timestamps you see in the first screenshot or integer timestamps you see in the second. You could try giving `--initial_prompt "Hello, welcome to my lecture."` to nudge the model to weigh more on the "with-punctuation mode".

**Reply by FurkanGozukara** [reply]
> Being an autoregressive model, Whisper has a certain chance to get stuck into a "no-punctuation mode". It also seems to be correlated with the tendency to create either precise timestamps you see in the first screenshot or integer timestamps you see in the second. You could try giving `--initial_prompt "Hello, welcome to my lecture."` to nudge the model to weigh more on the "with-punctuation mode".

Ty very much looks like working. I have given the command as like this 

`whisper "mvc lecture 1 watermarked v2.mp4" --model medium.en --language en --initial_prompt "Hello dear students. Welcome to the Lecture 1 of Responsive Web Design with ASP.NET 5 and MVC Pattern and Bootstrap 5. So, we will proceed this course with using C sharp language, programming language, and Microsoft SQL server."`

So now it will continue like this through whole processing? 

**Also it doesn't add initial prompt to the output file and timing? That would be huge problem.**

![image](https://user-images.githubusercontent.com/19240467/193146478-3f9881d7-bc14-4789-b542-b46fad41a39f.png)

**Reply by jongwook** [reply]
The model is trained to treat the prompt as something that came *before* the current audio. It appears that you gave the literal transcription of the first ~20 seconds, so the model thought the prompt was actually in the previous outputs and produced only the continuations in the current output (which you see in `00:00.000 --> 00:30.000`). It'll behave better if you prompt with a hypothetical sentence that may have come before the actual audio, like `"The following is a lecture on web development."`.

> So now it will continue like this through whole processing?

Likely so, but there's a chance it may switch to the no-punctuation mode again, like the reverse of what it did at the 75-minute mark originally.

**Reply by FurkanGozukara** [reply]
Ty so much. With hypothetical input like this it works now. But this is a really weird thing. Why a hypothetical sentence fixes the behaviour? Can't we force it all the time to generate in punctuation mode?

`--initial_prompt "Welcome to the Software Engineering Courses channel."`

![image](https://user-images.githubusercontent.com/19240467/193149106-55505e59-07d7-41b4-881a-254235cce827.png)

**Reply by jongwook** [reply]
It's a hacky but simplest interface to steer the model behavior. Welcome to the world of prompt engineering 😂. A more serious way for doing that would be to fine-tune the model with the transcripts with punctuations, but it's more cumbersome.

**Reply by FurkanGozukara** [reply]
> It's a hacky but simplest interface to steer the model behavior. Welcome to the world of prompt engineering 😂. A more serious way for doing that would be to fine-tune the model with the transcripts with punctuations, but it's more cumbersome.

Ye it got broken again after some time and got fixed again after some time. I wish there was a way to force it all the time do punctuation.

**Reply by ryanheise** [reply]
> Being an autoregressive model, Whisper has a certain chance to get stuck into a "no-punctuation mode". It also seems to be correlated with the tendency to create either precise timestamps you see in the first screenshot or integer timestamps you see in the second. You could try giving `--initial_prompt "Hello, welcome to my lecture."` to nudge the model to weigh more on the "with-punctuation mode".

That's really fascinating. It's almost as though Whisper has a split personality disorder :-)

I wonder, would it be possible to include non-integer timestamps in the prompt to nudge it toward accurate timestamps, and to also include punctuation in the prompt to get it to do both punctuation and accurate timestamps at the same time? Or would accurate timestamps tend to switch off the punctuation anyway?

**Reply by HaeChan0305** [reply]
@jongwook I'm struggled with same problem, no-punctuation mode. Why do you think that punctuation mode working is correlated with the tendency to create either precise timestamps?

**Reply by Sogl** [reply]
This solution helps for a while, and then it's back to the mode without punctuation and sentences.

**Reply by jonathanlal** [reply]
yeah doesn't work long term usage - some recordings seem to fail. It does retain some punctuation but I get unreasonably long sentences.

**Reply by rahulbansal16** [reply]
What is the solution? How did you got it fixed?

**Reply by ThioJoe** [reply]
I found using an initial prompt of just this:
```
Hello.
```

fixed the problem of no punctuation. I'm using WhisperCPP, not sure if that matters. So on the command line with WhisperCPP looks like this:
```
--prompt "Hello."
```

I mention that because I am not actually sure it's even stripping the quotes, so if the first one doesn't work maybe even try including the quotes in the literal string.

**Comment by FurkanGozukara** [comment]
@jongwook I am 100% there must be a way to force it to be on punctuation mode 100% time

Yesterday I did various tests and finally I made it stay

After converting audio into mp3, medium en model with a hypothetical sentence processed this video of mine with 100% punctuation 

https://www.youtube.com/watch?v=77iDUQd4x90

If only I knew Python I am sure I could figure out but I am C# guy

Here 2 transcribes files for you to compare what I mean

1 : MP3 formatted input, large model, with a hypothetical sentence, in the beginning punctuation on then it becomes off then becomes on again : https://docs.google.com/document/d/12lo_Utex7dpM1qLHnxYjsYTcLYzzYuCOLkbhP6KZH3U/edit?usp=sharing

2 : MP3 formatted input, medium en model, with a hypothetical sentence, 100% time punctuation : https://docs.google.com/document/d/1j1fTf_h-086mHHfCp74GbW2e6DIqo9GCqUlkKZAG0nQ/edit?usp=sharing

Both models got exactly same input and 1 of them worked with punctuation  on and another didn't

I really need help on this thank you very much

**Reply by brendankntb** [reply]
I have a similar situation. Transcribing with turbo model and the initial prompt and there is no punctuation. Using the exact same code changed to medium model produces a transcription with punctuation. I haven’t been able o find a way to get the turbo model to work reliably with punctuation.

**Comment by FurkanGozukara** [comment]
I also noticed very weird something. At some parts of the video, I make computer to read some text.

For example here at the minute 11:42 i start computer voice. As you can imagine it has perfect English. Just open subtitles it is generated by Whisper.

https://youtu.be/77iDUQd4x90?t=702

However, model medium.en didn't generate any text for that part. It generated almost perfect text for my speech but that part is missing.

This behaviour repeats. Sometimes entire speech of computer is missing and sometimes only some part of it.

I have used this command to generate this transcription

`whisper input.mp3 --model medium.en --language en --initial_prompt "Welcome to the Software Engineering Courses channel."`

@jongwook

**Reply by jongwook** [reply]
The model sometimes gets confused by a sudden change of voice and may drop the second speaker's part due to the training data that didn't contain captions for interviewees or chatter in the background. There is unfortunately no surefire way to do this, but you might get a different behavior by adjusting the beam size (to have more candidates which is likely to contain the second speaker) and length penalty (to penalize less on longer transcription), or even by giving `--suppress_tokens ""` which may output unwanted symbols but possibly be more likely to transcribe the second speaker.

**Reply by FurkanGozukara** [reply]
Second speaker is speaking alone in my video. So either speaker 1 speaks or speaker 2 speaks. I will test --suppress_tokens "" and beam size. What beam size you suggest like 10 15 20? thank you very much. also when we run default cmd it is 5 right?

**Reply by jongwook** [reply]
> So either speaker 1 speaks or speaker 2 speaks.

Right, even if they are not simultaneously speaking, the model might get offguarded and (wrongfully) decide to drop the second speaker.

> What beam size you suggest like 10 15 20?

I know it sounds disappointing but again there's no correct answer on this; we observed for some audio larger beam size is better, for some 5 worked best and it began to degrade for larger beam sizes.

> when we run default cmd it is 5 right?

Yes

**Reply by FurkanGozukara** [reply]
Dear @jongwook , could be that --suppress_tokens "" causing a bug? Because now medium model uses much more ram and it generated 0 output 🗡️ 

![image](https://user-images.githubusercontent.com/19240467/193360351-f3db831e-f2e7-42ed-90ec-64187188f293.png)

edit : beam size 10 captured that missing part above. still processing. i will compare once it is done. ty

edit 2 : beam 10 seriously decreased punctuation quality. so not useful for me :(

**Comment by FurkanGozukara** [comment]
Perhaps we could implement this into the whisper optionally? That can process output of whisper and save as another output?

https://github.com/xashru/punctuation-restoration

**Comment by ANonEntity** [comment]
I suspect the reason it drops the punctuation is this:
https://github.com/openai/whisper/blob/0b1ba3d46ebf7fe6f953acfd8cad62a4f851b49f/whisper/transcribe.py#L235-L237
The model struggles with a segment, resets the prompt, and then decides to go without punctuation from there. I've made a pull request (#220) that I think might solve it.

**Reply by FurkanGozukara** [reply]
> I suspect the reason it drops the punctuation is this:
> 
> https://github.com/openai/whisper/blob/0b1ba3d46ebf7fe6f953acfd8cad62a4f851b49f/whisper/transcribe.py#L235-L237
> 
> 
> The model struggles with a segment, resets the prompt, and then decides to go without punctuation from there. I've made a pull request (#220) that I think might solve it.

Awesome. I am gonna test it now to see.

**Reply by FurkanGozukara** [reply]
> I suspect the reason it drops the punctuation is this:
> 
> https://github.com/openai/whisper/blob/0b1ba3d46ebf7fe6f953acfd8cad62a4f851b49f/whisper/transcribe.py#L235-L237
> 
> 
> The model struggles with a segment, resets the prompt, and then decides to go without punctuation from there. I've made a pull request (#220) that I think might solve it.

unfortunately it failed for me :(

![image](https://user-images.githubusercontent.com/19240467/193430166-dedc1ab1-a973-4cd4-90db-dec89543ee89.png)

![image](https://user-images.githubusercontent.com/19240467/193430186-d7f8ea5d-769d-4dc1-94bc-c1b8e339763b.png)

**Reply by ANonEntity** [reply]
No, that's not the fix. You need to add these two lines after "decode_options["prompt"] = all_tokens[prompt_reset_since:]":
https://github.com/openai/whisper/blob/34f971e661e3de17c3576b7839304463013e395c/whisper/transcribe.py#L178-L179

**Reply by FurkanGozukara** [reply]
> No, that's not the fix. You need to add these two lines after "decode_options["prompt"] = all_tokens[prompt_reset_since:]":
> 
> https://github.com/openai/whisper/blob/34f971e661e3de17c3576b7839304463013e395c/whisper/transcribe.py#L178-L179

I copy pasted entire file from your link and still not working

here the transcribe file i am using : 

[transcribe.txt](https://github.com/openai/whisper/files/9691340/transcribe.txt)

**Reply by ANonEntity** [reply]
Looks like that's not the issue, then. Oh well 🙁

**Reply by FurkanGozukara** [reply]
> Looks like that's not the issue, then. Oh well 🙁

here the mp3 file i am testing : https://drive.google.com/file/d/1HrT_CnZzifQ9JFihGLWk7YQo0WvEO-Wq/view?usp=sharing

and this the command : whisper --model medium.en --lang en

it is broken after like 40 minutes. actually it is mixed broken and mixed working

![image](https://user-images.githubusercontent.com/19240467/193431653-6d73310b-b923-4efa-9c22-c2b8ba2372fa.png)

it also has some glitch 

![image](https://user-images.githubusercontent.com/19240467/193431657-1d12c618-0f12-4302-b2f5-b8fd2a57528a.png)

**Comment by gudh** [comment]
Sorry to interrupt

But is it possible for `whisper` to punctate the transcript alone?

I mean, if I downloaded the auto-generated subtitles from a youtube video with `youtube-dl`, is it possible to punctate the subtitles with `whisper`?

**Reply by jongwook** [reply]
Not out of the box. If the two transcripts are sufficiently similar, you could match them using an [edit distance](https://en.wikipedia.org/wiki/Edit_distance) metric and apply punctuations. Alternatively, you could try a language model like GPT-3 to do the job without having to deal with the audio:

<img width="747" alt="image" src="https://user-images.githubusercontent.com/266841/196248133-6894fba9-a99c-443b-be62-a7c9fe5c85e5.png">

**Reply by FurkanGozukara** [reply]
only if it was free :(

**Reply by rahulbansal16** [reply]
@FurkanGozukara @jongwook the problem with this approach is sometimes it messes up intricacies of user pause for example


Input Sentence: "i want to store these logs at my end also. Is there a way to do that I don't want to add latency at my end so is there a way i can use some webhook after you are done with the processing"


Output Sentence "I want to store these logs at my end also. Is there a way to do that? I don't want to add latency at my end, so is there a way I can use some webhook after you are done with the processing?"

Output Sentence 2 "I want to store this log at my end also. What is the way to do that? I don't want to add latency at my end, so is there a way I can use some webhook? After you're done with the processing.,"

**Comment by mayeaux** [comment]
Strangely enough, if I pass an initial prompt of `Hello.` for English content, it fixes previously punctuation-less content even if the word 'Hello' never appears. It seems like with some content the ASR recognizes it as 'crazy punctuation free content' (happens to me with content that jumps around in the narration or has bad quality), perhaps some of the training data was punctuation free like that and it determines that it's 'one of those types of content'. Adding the `Hello.` seems to indicate to Whisper that actually this isn't a crazy punctuation free piece of content it's a piece of content that should be punctuated.

**Reply by otakutyrant** [reply]
Now `Hello` become a myterial dark magic that worked for me.

**Reply by ThioJoe** [reply]
Found this thread through google, but the `Hello.` initial prompt seemed to work for me as well.

**Comment by mayeaux** [comment]
This is a hack solution I came up with today:

https://github.com/mayeaux/faster-whisper/commit/dda1795cfffe09d819fb94b4c8bf78756362bc6e

It could use some refining but honestly it works well for me.

**Reply by 4drawing95** [reply]
I have a question. My coding skills are pretty limited, so I'm not sure. How do you reflect that update? I modified that translate.py, but it only increases the printed output, no change in the actual output. do I need to do something extra?

**Reply by Saccarab** [reply]
This works fairly well when condition on prev text is enabled however, I've come across cases like below where when whisper actually decides to hallucinate these supplied tokens punctuations show themselves over and over again. so overall seems to be a little problematic still.

![image](https://github.com/openai/whisper/assets/15700171/c4b5f191-eb2f-4432-9546-e2cdd63cccd3)

**Comment by FurkanGozukara** [comment]
I started using GPT4 to fix punctuation :D

**Reply by ejentos** [reply]
@FurkanGozukara do you use GPT4 whisper or adjust already transcribed audio without punctuation and ask GTP4 to add it?

**Reply by FurkanGozukara** [reply]
> @FurkanGozukara do you use GPT4 whisper or adjust already transcribed audio without punctuation and ask GTP4 to add it?

i transcribe first with Whisper
then i fix transcription

**Reply by jonathanlal** [reply]
How would you align your subtitles then if they have been edited or changed by post processing?

**Reply by rahulbansal16** [reply]
@FurkanGozukara  the problem with this approach is sometimes it messes up intricacies of user pause for example

Input Sentence: "i want to store these logs at my end also. Is there a way to do that I don't want to add latency at my end so is there a way i can use some webhook after you are done with the processing"

Output Sentence "I want to store these logs at my end also. Is there a way to do that? I don't want to add latency at my end, so is there a way I can use some webhook after you are done with the processing?"

Output Sentence 2 "I want to store this log at my end also. What is the way to do that? I don't want to add latency at my end, so is there a way I can use some webhook? After you're done with the processing.,"

**Comment by sjtu-hxj** [comment]
This maybe works: `--initial_prompt "Please do not forget the punctuations!"`

**Comment by Zigunov** [comment]
Hey all! I found that my “—initial_prompt” would work for a short time. And it would stop after a while. I found that adjusting the prompt by adding in 1 more sentence or word that has punctuation or full stop managed to circumvent the repetitive failure loop of getting the same transcript. It would almost be an entirely new transcript. 

Hope this helps!

**Reply by Zigunov** [reply]
Update on this again, I’ve had a very high success rate by making the initial prompt be text taken from a text file called “initialprompt.txt” and I’ve automated it so that anytime there’s a failure, it takes a new sentence from a “sentence bank” text file (with lots of various sentences with punctuation, taken from previous transcripts) and adds it to the initial prompt text file to add variety for whisper to look at the new initial prompt as if it is brand new! Then it re-transcribes and so far, I’ve had 100% success with this tactic.

**Comment by Zigunov** [comment]
I found that if the prompt stops working well, to just add a word with
punctuation and that should fix the issue. Every time it stops just add
something to the prompt.

**Reply by rahulbansal16** [reply]
@Zigunov can you elaborate on it?

**Comment by Zigunov** [comment]
So for the initial prompt, writing “Hello.” Into a text file and
referencing it in the whisper script as usual. But if you find that it
stops functioning, go into that text file and add maybe a sentence after
Hello, like “Goodbye, my friend.” Which should fix the issue, because it’s
a new initial prompt.

**Reply by rahulbansal16** [reply]
I used the whisper to determine the language also. So, what are the alternatives that I have? Does putting the english prompt might mess up with the non english language?

**Comment by Zigunov** [comment]
I would say try both options. Whisper is very fast now, especially Whisper
XXL. Are you quality checking them yourself?

**Reply by rahulbansal16** [reply]
I am running it in my product and can't figure it out myself.
